File: langs/python/src/vmx/messages/subscribe_value.py
"""Imperative selected-state bridge over a fixed VM's message hub."""

from __future__ import annotations

import operator
from collections.abc import Callable
from typing import TYPE_CHECKING, TypeVar

from reactivex.abc import DisposableBase

from vmx.messages.property_changed import PropertyChangedMessage

if TYPE_CHECKING:
    from vmx.components.protocols import ComponentVMProto

TSource = TypeVar("TSource", bound="ComponentVMProto")
TValue = TypeVar("TValue")


def subscribe_value(
    source: TSource,
    selector: Callable[[TSource], TValue],
    callback: Callable[[TValue, TValue], None],
    *,
    equality: Callable[[TValue, TValue], bool] | None = None,
    fire_immediately: bool = False,
) -> DisposableBase:
    """Subscribe an imperative callback to selected state from one fixed VM."""
    current = selector(source)
    if fire_immediately:
        callback(current, current)
    equality_fn: Callable[[TValue, TValue], bool] = equality or operator.eq

    def on_message(message: object) -> None:
        nonlocal current
        if not isinstance(message, PropertyChangedMessage) or message.sender is not source:
            return
        next_value = selector(source)
        if equality_fn(current, next_value):
            return
        previous = current
        current = next_value
        callback(next_value, previous)

    return source.hub.messages.subscribe(on_message)

File: langs/typescript/src/messages/subscribeValue.ts
import type { Subscription } from "rxjs";
import type { IComponentVM } from "../components/types.js";
import { isPropertyChanged } from "./predicates.js";

export interface SubscribeValueOptions<TValue> {
  readonly equality?: (current: TValue, next: TValue) => boolean;
  readonly fireImmediately?: boolean;
}

export function subscribeValue<TSource extends IComponentVM, TValue>(
  source: TSource,
  selector: (source: TSource) => TValue,
  callback: (current: TValue, previous: TValue) => void,
  options?: SubscribeValueOptions<TValue>,
): Subscription {
  let current = selector(source);
  if (options?.fireImmediately === true) callback(current, current);
  const equality = options?.equality ?? Object.is;

  return source.hub.messages.subscribe((message) => {
    if (!isPropertyChanged(message, { sender: source })) return;
    const next = selector(source);
    if (equality(current, next)) return;
    const previous = current;
    current = next;
    callback(next, previous);
  });
}
